### PR TITLE
Introduce/extend `RequireNonNullElse{,Get}` Refaster rules

### PR DESCRIPTION
**Pull request by benhalasi**
Related issue: #364

Added Rules:
 - prefer `requireNonNullElse(A, B)` over `Optional.ofNullable(A).orElse(B)`
 - prefer `requireNonNullElseGet(A, () -> B)` over `Optional.ofNullable(A).orElseGet(() -> B)`

 Wrote test inputs and outputs.

**Potential problems:**

In the tests, referenced `Optional` by it's fully qualified name to avoid breaking `RefasterRulesTest` or leaving unused imports in the code. - this seems hacky but didn't came up with better solution.

Neither the new or already existing rules are not dealing with potential NPE when the second argument can be or produce null. - this seems to be a larger or at least a separated issue.

### CODE REVIEW

**Comment by Stephan202** [comment]
Thanks for the PR @benhalasi! I'll have a closer look at the changes later.

> In the tests, referenced `Optional` by it's fully qualified name to avoid breaking `RefasterRulesTest` or leaving unused imports in the code. - this seems hacky but didn't came up with better solution.

Yes, this a known issue. For this we have a little trick; see how [`elidedTypesAndStaticImports`](https://github.com/PicnicSupermarket/error-prone-support/search?q=elidedTypesAndStaticImports) is used :).

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/NullRules.java
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/NullRules.java]: This kind of subtype matching is done automatically by Refaster. That said, there's indeed work to be done here: we should introduce `S extends T`, so that we can match ` Supplier<S>`.

**Review by Stephan202 — APPROVED** [review]
Rebased and added a commit. Thanks for your contribution @benhalasi!

Suggested commit message:
```
Introduce/extend `RequireNonNullElse{,Get}` Refaster rules (#425)
```

The next release may be a few weeks out, but if you'd like to use these changes sooner, let us know and we'll see what we can do :)

**Review by rickie — APPROVED** [review]
Rebased and have one small remark for @Stephan202. 

Aside from that; changes LGTM! Suggested commit message also LGTM!

@benhalasi Thanks for your contribution!

**Comment by rickie** [comment]
Shall we add `Fixes #364` to the suggested commit message and create a new ticket for the specific rewrite you mention there @Stephan202 : `requireNonNullElse(map.get(k), defaultValue) -> map.getOrDefault(k, defaultValue)`?

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/NullRules.java
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/NullRules.java:54]: `s/master/matcher/` ;)

Indeed I started work on that [here](https://github.com/PicnicSupermarket/error-prone-support/compare/master...sschroevers/introduce-IsNullableMatcher), but I recall hitting some issues. (The `AbstractMatcherTestChecker` changes are part of that, but IIRC there was more. Would have to check out the branch and run the tests to refresh my memory.)

**Comment by Stephan202** [comment]
> Shall we add `Fixes #364` to the suggested commit message and create a new ticket for the specific rewrite you mention there @Stephan202 : `requireNonNullElse(map.get(k), defaultValue) -> map.getOrDefault(k, defaultValue)`?

SGTM. (There are also @EnricSala's observations around `Optional.of`, which are perhaps worth tracking separately? Those may require a `BugChecker`, because we'll likely want two suggested fixes: (a) changing `of` to `ofNullable` and (b) using `requireNonNull` instead.)

**Comment by rickie** [comment]
Rebased.
 
Will soon make the earlier mentioned GH issues. Sorry, haven't come around do doing it yet. Will merge right after that.

**Comment by Stephan202** [comment]
The follow-up issues are #431 and #436.

**Pull request merged**
